glommio/src/sync/gate.rs

```diff
@@ -0,0 +1,188 @@
+use std::cell::{Cell, RefCell};
+use std::rc::Rc;
+
+use futures_lite::Future;
+
+use crate::channels::local_channel::{self, LocalSender};
+use crate::{GlommioError, Local, ResourceType, Task, TaskQueueHandle};
+
+#[derive(Debug)]
+enum State {
+    Closing(LocalSender<bool>),
+    Closed,
+    Open,
+}
+
+/// Facility to achieve graceful shutdown by waiting for the dependent tasks
+/// to complete.
+#[derive(Clone, Debug)]
+pub struct Gate {
+    inner: Rc<GateInner>,
+}
+
+impl Gate {
+    /// Create a new [`Gate`](crate::sync::Gate)
+    #[allow(clippy::new_without_default)]
```

Comment:
why not add a `Default` ?
The new implementation looks like what a good default would be. We can probably implement `Default` on `State` (I think it is a good practice to implement `Default` for enums, since you can't derive it, and then likely just `Derive` it for `Gate`.

I will not consider this a blocker because the code is very good otherwise, but please consider sending a follow up

Reply:
IMO, people would always create `Gate`s explicitly with `new`. Is there a case in which a default value of `Gate` would be used via the `Default` trait?

Reply:
There are a variety of functions in rust that expect a default implementation, and may be more convenient to use.
For example, `std::mem::take`

If something can be thought to have an implementation that is a reasonable default, I think we should expose  it.

Reply:
There's a risk of the default `Gate` created by `std::mem::take` being used unconscious, and left unclosed at shutdown. If it's real necessary to take an old `Gate` and leave a new one in-place, then `std::mem::replace(&gate, Gate::new())` should be better used to express the users intention explicitly, IMO.

Reply:
The risk of leaving the gate unclosed is still there in your example, if the programmer forgets to close after replacing. Sure, it is a better pattern and it is the one I expect to see, but if we are going to force a specific pattern I would expect that pattern to be really fool proof, and this isn't. At least not enough.

Consider the downsides:

Imagine a user has a temporary struct with 100 members, one of them being a gate. User wants to add `[derive(Default)]` and is now unable to as `Gate` doesn't have a default implementation. Is the guarantee given by forbidding `mem::take` enough to justify this hassle? I think that's the question that really matters

Reply:
> Imagine a user has a temporary struct with 100 members, one of them being a gate. User wants to add `[derive(Default)]` and is now unable to as `Gate` doesn't have a default implementation. Is the guarantee given by forbidding `mem::take` enough to justify this hassle? I think that's the question that really matters

I really recommend a refactoring to split the struct into smaller pieces in this circumstance and move the gate member to a part where is `Default` is not necessary. If not, a slightly verbose implementation of `Default` could be done, not as simple as `[derive(Default)]` though: 
```rust
impl Default for MyStruct {
    fn default() -> Self {
        Self { gate: Gate::new(), ..Defaults::default() }
    }
}
```

Anyway, I'm not purticularly against the idea of making the `Gate` a `Default`. I'm just not convinced it would be necessary in a real project (not imaginary) with proper design. I think it could be done when a user send a request for it.